### DISCUSSION TITLE
"brew tap user/repo" fails both with correct Github credentials and also when aborting U/P input

### DISCUSSION THREAD

**Discussion by porg** [Homebrew/discussions#1074] [Getting started]
# Environment
``` zsh
➜  ~ sw_vers
ProductName:	macOS
ProductVersion:	11.2.1
BuildVersion:	20D74
➜  ~ brew --version
Homebrew 3.0.7
Homebrew/homebrew-core (git revision d44a2e2271; last commit 2021-03-18)
Homebrew/homebrew-cask (git revision 2220b59833; last commit 2021-03-18)
```

# I ran these before
`brew upgrade`	➜ Upgraded some packages. No errors.
`brew doctor`	➜ Nothing significant or potentially causally related IMHO. If relevant see [brew-doctor.txt](https://github.com/Homebrew/discussions/files/6166524/brew-doctor.txt)

# When I "brew tap" a certain Github user/repo if fails
```zsh
➜  ~ brew tap ross-spencer/zbthumbnail-extractor
==> Tapping ross-spencer/zbthumbnail-extractor
Cloning into '/usr/local/Homebrew/Library/Taps/ross-spencer/homebrew-zbthumbnail-extractor'...
Username for 'https://github.com': porg
Password for 'https://porg@github.com': # Entering exactly as I did in the webbrowser where it works!
remote: Repository not found.
fatal: repository 'https://github.com/ross-spencer/homebrew-zbthumbnail-extractor/' not found
Error: Failure while executing; `git clone https://github.com/ross-spencer/homebrew-zbthumbnail-extractor /usr/local/Homebrew/Library/Taps/ross-spencer/homebrew-zbthumbnail-extractor --origin=origin` exited with 128.
```

# Gave it another try
This time pressing ctrl-D when asked for user/password:
```zsh
➜  ~ brew tap ross-spencer/zbthumbnail-extractor
==> Tapping ross-spencer/zbthumbnail-extractor
Cloning into '/usr/local/Homebrew/Library/Taps/ross-spencer/homebrew-zbthumbnail-extractor'...
Username for 'https://github.com': fatal: could not read Username for 'https://github.com': Undefined error: 0
Error: Failure while executing; `git clone https://github.com/ross-spencer/homebrew-zbthumbnail-extractor /usr/local/Homebrew/Library/Taps/ross-spencer/homebrew-zbthumbnail-extractor --origin=origin` exited with 128.
➜  ~ 
```

# Research
1. Before submitting this as a bug, could someone please try to reproduce this?
2. Possibly related issues I found:
  - https://github.com/Homebrew/legacy-homebrew/issues/18432
  - https://github.com/Homebrew/brew/issues/10766

**Comment by carlocab** [comment]
https://github.com/ross-spencer/zbthumbnail-extractor doesn't appear to be a Homebrew tap. That won't stop you from tapping it if you really wanted to (I think), but it's not going to do anything but clone the repository into a Homebrew-specific location.

However, if you're still convinced you want to tap it, try
```
brew tap ross-spencer/zbthumbnail-extractor https://github.com/ross-spencer/zbthumbnail-extractor.git
```

**Reply by porg** [reply]
> doesn't appear to be a Homebrew tap

Naive me!
I thought that `brew tap` can **download arbitrary Github repos** (which it can actually as you indicate)
**plus make them work** then with some intelligent algorithms (which was my fairy tale dream):
- from **manifest-files** keeps the files in the Cellar under some versioned folders
- looks for a **make-file** (or similar build instructions)
- and then builds the executables and creates the necessary opt-links

But if you just think about it longer, you realize:
- Cannot be just any git repo.
- The repo needs to fulfill certain criteria and provide certain structures so that a package manager can work with it.

# What do we learn from this? What can be improved?
1. I had prepared myself with reading the manpage.
--> `man brew`: The whole section on `tap` looses not a single word on the circumstance that in order to be compatible a Github project needs to fulfill certain criteria, and which criteria that are. (What manifest files or precompiled files it looks for, etc)
2. And I ran into the error of attempting to `brew tap` a repo which is not `brew tap`compatible!
--> A more specific and helpful error message would help which mentions that you can not simply tap any repo but that it must fulfill certain criteria! And a reference to a URL or manpage/section to what these criteria are.

These two improvements would avoid the pitfall that I fell into!

**Reply by carlocab** [reply]
This page might help, as it helps clarify some of your initial confusion: https://docs.brew.sh/Taps

See also:

https://docs.brew.sh/Interesting-Taps-and-Forks
https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

PRs to improve the docs welcome.

**Reply by carlocab** [reply]
Note that `man brew` actually already says it expects `homebrew` to appear in the tap name by default:
```
Usage: brew tap [options] [user/repo] [URL]

Tap a formula repository.

If no arguments are provided, list all installed taps.

With URL unspecified, tap a formula repository from GitHub using HTTPS. Since
so many taps are hosted on GitHub, this command is a shortcut for brew tap
user/repo https://github.com/user/homebrew-repo.

With URL specified, tap a formula repository from anywhere, using any
transport protocol that git(1) handles. The one-argument form of tap
simplifies but also limits. This two-argument command makes no assumptions, so
taps can be cloned from places other than GitHub and using protocols other than
HTTPS, e.g. SSH, git, HTTP, FTP(S), rsync.
```

**Reply by porg** [reply]
> With URL unspecified, tap a formula repository from GitHub using HTTPS. Since so many taps are hosted on GitHub, this command is a shortcut for brew tap user/repo https://github.com/user/homebrew-repo.

This is a golden example of very technical writing, which may be specific and correct and whatnot, but implicitly requires prior knowledge instead of being just plain and presenting the idea itself.

As a UX designer, copy text is part of my job, so I instead of just moaning, I provide a preliminary improved wording:

----

Usage: brew tap [options] [user/repo] [URL]

Tap a formula repository.

That is a homebrew compatible repository containing all what's necessary for brew to fetch and install it. By convention they start with a "homebrew-" prefix.

If no arguments are provided, list all installed taps.

Not stating [URL] but simply [user/repo] assumes this is a tap hosted on Github, as are the majority of taps. This short form gets expanded to a full canonical Github URL in the form of https://github.com/user/homebrew-repo. Note that in the expansion the prefix "homebrew-" gets added before "repo" as is the convention for taps. If your desired tap does not exist at Github with the "homebrew-" prefix, it is very likely not homebrew compatible, see: https://docs.brew.sh/Taps

With URL specified […]

**Reply by carlocab** [reply]
Feel free to open a PR to fix this. The place this needs to be modified is here:

https://github.com/Homebrew/brew/blob/1b61d5a5639d7cf76a1fc3faf02c87f0aef9e64c/Library/Homebrew/cmd/tap.rb#L15-L29

Note, however, that these descriptions are deliberately terse, so you may wish to cut the wording down a little.
